xapian-core/cluster/cluster.cc

```diff
@@ -246,12 +236,15 @@ Point::initialize(TermListGroup &tlg, const Document &doc_) {
     for (TermIterator it = doc.termlist_begin(); it != doc.termlist_end(); ++it) {
 	doccount wdf = it.get_wdf();
 	string term = *it;
-	if (wdf < 1 || term[0]=='Z')
+	double termfreq = tlg.get_termfreq(term);
+	/** If term is not a stemmed term or the term indexes only one
+	 *  one document, do not compute TF-IDF scores for them
+	 */
+	if (wdf < 1 || term[0] != 'Z' || termfreq <= 1)
```

Comment:
Not easily, because we'd have to configure clusterers with `Point` subclasses, which you can't do directly (we'd have to provide an object which could be cloned or something, which would likely result in an unnatural interface).

Perhaps we should take a step back and think about what we're trying to do here. We want to take an `MSet` and generate `Point` objects each of which represents a document in the (term-based) vector space (plus some cached attributes). So I wonder if we should keep `Point` more basic, and change `TermListGroup` to be a `Point`-emitting object that works from an `MSet`, instead of having `Point::initialize()` build the vector.

So we'd replace the `FreqSource` hierarchy with a `PointSource`, something like this:

```c++
vector<Point> points;
MSetPointSource ps(enq.get_mset(0, 10));

for (Xapian::PointIterator pit = ps.points_begin();
     pit != ps.points_end(); ++pit) {
    points.push_back(*pit);
}
```

Then, we could pass a pair of `PointIterator` objects into `Clusterer::cluster()`, instead of the `MSet` itself. That means that the user code controls the `PointSource`, so it can be subclassed for different filtering. The source would have to read in all the documents on initialisation, and then could filter things out as you use the iterator. (So basically the vector construction in `Point::initialize()` would be done in the iterator and then passed to the `Point` constructor.)

@ojwb do you have thoughts on this? I certainly feel that having an unchangeable filter in `Point` is the wrong approach, and it'll only get worse as we add things like stop list support and other dimensionality reduction to the vector generation.

I know that originally there was some thought about other sources of term frequencies that could be used to build points, but until we have a concrete example we want to implement we shouldn't be swayed by that.